File: Linked lists/linked_list.py

class Node:
    def __init__(self, data) -> None:
        self.data = data
        self.next = None


class LinkedList:

    def __init__(self) -> None:
        self.head = None

    def insertBeg(self, data):
        # to insert at beggining
        newNode = Node(data)
        newNode.next = self.head
        self.head = newNode

    def insertLast(self, data):
        # to insert at end
        if self.head == None:
            self.head = Node(data)

        temp = self.head
        while (temp.next is not None):
            temp = temp.next

        temp.next = Node(data)

    def printList(self):
        temp = self.head
        while temp is not None:
            print(temp.data, end=" ")
            temp = temp.next
        print()
    
    def deleteNode(self,data):
        if self.head is None:
            return
        
        if self.head.data==data:
            self.head=None
        
        temp=self.head
        prev=None
        while  temp.data !=data and temp.next is not None :
            prev=temp
            temp=temp.next
        
        if temp is None:
            print("not found")
        else:
            prev.next=temp.next
              
        

if __name__ == "__main__":
    list=LinkedList()
    list.insertBeg(1)
    list.insertBeg(2)
    list.insertLast(3)
    list.insertLast(5)
    list.deleteNode(3)

    list.printList()
    
    

File: Linked lists/linked_list.cpp
#include <iostream>

template <typename T>

struct node {
  node* next;
  T data;
};

template <typename T>
class LinkedList {
 private:
  node<T>* head;

 public:
  LinkedList() { head = nullptr; }

  void insertBeg(T value) {  // to insert a beggining of a linked list
    node<T>* newNode = new node<T>{nullptr, value};
    newNode->next = head;
    head = newNode;
  }

  void insertLasr(T value) {  // to onsert at end of list
    node<T>* newNode = new node<T>(value);
    if (head == nullptr) {
      head = newNode;
      return;
    }
    node<T>* temp = head;
    while (temp->next != nullptr) {
      temp = temp->next;
    }
    temp->next = newNode;
  }

  void deleteNode(T value) {
    if (head == nullptr) {
      return;
    }
    if (head->data == value) {
      node<T>* temp = head;
      head = head->next;
      delete temp;
      return;
    }
    node<T>* temp = head;
    while (temp->next != nullptr && temp->next->data != value) {
      temp = temp->next;
    }
    if (temp->next == nullptr) {
      // node<T> with the given value not found
      return;
    }
    node<T>* nodeToDelete = temp->next;
    temp->next = temp->next->next;
    delete nodeToDelete;
  }

  void printList() {
    node<T>* temp = head;
    while (temp != nullptr) {
      std::cout << temp->data << " ";
      temp = temp->next;
    }
    std::cout << std::endl;
  }
};

int main(void){

LinkedList<int> list;
 list.insertBeg(1);
 list.insertBeg(1);
 list.insertBeg(1);
 list.insertBeg(1);

 list.printList();

 list.deleteNode(1);

 list.printList();



}
